optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/ConsecutiveIntervalInfoImpl.java

```diff
@@ -0,0 +1,214 @@
+            clusterStartSplitPointToCluster.put(start, newCluster);
+            Map.Entry<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> nextClusterEntry =
+                    clusterStartSplitPointToCluster.higherEntry(start);
+            if (nextClusterEntry != null) {
+                clusterStartSplitPointToNextBreak.put(start,
+                        new IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>(newCluster,
+                                nextClusterEntry.getValue(), differenceFunction.apply(
+                                        newCluster.getEnd(), nextClusterEntry.getValue().getStart())));
+            }
+            Map.Entry<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> previousClusterEntry =
+                    clusterStartSplitPointToCluster.lowerEntry(start);
+            if (previousClusterEntry != null) {
+                clusterStartSplitPointToNextBreak.put(previousClusterEntry.getKey(),
+                        new IntervalBreakImpl<>(previousClusterEntry.getValue(), newCluster, differenceFunction.apply(
+                                previousClusterEntry.getValue().getEnd(), newCluster.getStart())));
+            }
+            return;
+        }
+        NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>> intersectedIntervalBreakMap =
+                clusterStartSplitPointToNextBreak.subMap(
+                        ObjectUtils.defaultIfNull(clusterStartSplitPointToNextBreak.floorKey(interval.getStartSplitPoint()),
+                                interval.getStartSplitPoint()),
+                        true, interval.getEndSplitPoint(), true);
+        IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_> intervalCluster =
+                intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey());
+        IntervalSplitPoint<IntervalType_, PointType_> oldStart = intervalCluster.getStartSplitPoint();
+        intervalCluster.addInterval(interval);
+        intersectedIntervalClusterMap.tailMap(intersectedIntervalClusterMap.firstKey(), false).values()
+                .forEach(intervalCluster::mergeIntervalCluster);
+        intersectedIntervalClusterMap.tailMap(intersectedIntervalClusterMap.firstKey(), false).clear();
+
+        if (intersectedIntervalBreakMap.size() > 0) {
+            if (interval.getStartSplitPoint().splitPoint
+                    .compareTo(intersectedIntervalBreakMap.firstEntry().getValue().getPreviousIntervalClusterEnd()) <= 0) {
+                if (interval.getEndSplitPoint().splitPoint
+                        .compareTo(intersectedIntervalBreakMap.lastEntry().getValue().getNextIntervalClusterStart()) >= 0) {
+                    intersectedIntervalBreakMap.clear();
```

Comment:
Three nested ifs with non-trivial conditions.
In this case, I would consider making the bodies of those statements their own methods. Break the flow, encapsulate, help people to not have to think about so many things at once.